python/ray/data/dataset_pipeline.py

```diff
@@ -179,11 +180,25 @@ def iter_batches(
         )
         self._stats.iter_total_s.add(time.perf_counter() - time_start)
 
-    def _iter_blocks(self) -> Iterator[Block]:
+    def _iter_blocks(self, prefetch_blocks: int = 0) -> Iterator[Block]:
         ds_wait_start = time.perf_counter()
+        latest_blocks = collections.deque(maxlen=prefetch_blocks + 1)
```

Comment:
Would it be easier to delete blocks in the batch_blocks / iter_batches method instead of in _iter_blocks? Seems like that avoid avoid needing this deque.

Reply:
+1, I think that this could be made part of the prefetching logic with a small adjustment to the [sliding window implementation](https://github.com/ray-project/ray/blob/a9a81f3d93bdecbce445b3fce0866a083d3b9557/python/ray/data/impl/block_batching.py#L118-L143) and use. E.g., move your deque length check + object ref free into the sliding window implementation.

Reply:
I was thinking this should keep the change narrow and not to affect dataset.iter_batches() since batch_blocks() is shared. But I think it'll be good to see the laziness of dataset and datasetpipeline in a more unified way, so I moved the change to batch_blocks() now.

Reply:
> Could we add a unit test somehow for this (maybe pick a config that would hang/crash, or test the throughput)

Added.